encodings/dict/src/compute/cast.rs

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use vortex_array::compute::{CastKernel, CastKernelAdapter, cast};
+use vortex_array::{ArrayRef, IntoArray, register_kernel};
+use vortex_dtype::DType;
+use vortex_error::VortexResult;
+
+use crate::{DictArray, DictVTable};
+
+impl CastKernel for DictVTable {
+    fn cast(&self, array: &DictArray, dtype: &DType) -> VortexResult<Option<ArrayRef>> {
+        // Cast the dictionary values to the target type
+        let casted_values = cast(array.values(), dtype)?;
+
+        // Create a new dictionary array with the same codes but casted values
+        Ok(Some(
+            DictArray::try_new(array.codes().clone(), casted_values)?.into_array(),
+        ))
```

Comment:
until we support mixed nullability dict children you have to also cast the codes

Reply:
Basically if you write a test that casts AllValid dict to NonNullable or back this will fail